### ISSUE TITLE
Mapping pojo with lombok builder and usage of @BeforeMapping leads to compile error

### ISSUE THREAD

**Issue by Brutus5000** [mapstruct/mapstruct#2057]
In my project I am using the CycleAvoidingMappingContext by Andreas Gudian, which is basically an object cache for mapping. This worked perfectly until I recently moved all my dtos to use lomboks `@Builder` (or `@Superbuilder` sometimes).

I stripped the error down to a minimal example here: https://github.com/Brutus5000/mapstruct-lombok-beforemapping.

This is the from the context used in my Mapper:
```
    @BeforeMapping
    public <T> T getMappedInstance(Object source, @TargetType Class<T> targetType) {
        return (T) knownInstances.get(source);
    }
```

This is generated mapping code:
```
    @Override
    public BuilderDtoTarget map(BuilderDto object) {
        BuilderDtoTarget target = cycleAvoidingMappingContext.getMappedInstance( object, BuilderDtoTargetBuilder.class );
        if ( target != null ) {
            return target;
        }

        if ( object == null ) {
            return null;
        }

        BuilderDtoTargetBuilder builderDtoTarget = BuilderDtoTarget.builder();

        cycleAvoidingMappingContext.storeMappedInstance( object, builderDtoTarget );

        builderDtoTarget.id( object.getId() );

        return builderDtoTarget.build();
    }
```

As you can see it tries to assign the variable `BuilderDtoTarget target` with a type of its own builder class.

What am I doing wrong here? Is this a bug? I'm not entirely sure.

**Comment by sjaakd** [comment]
The modifiable thing in a builder constructed class is the builder itself. In many use cases, builders are used to construct immutable datatypes. Hence, builders are passed to `@BeforeMapping` `@AfterMapping`. 

Is it absolutely necessary to use a builder here? Otherwise my advice would be to stick to a getter/setter dto or to disable the builder (MapStruct 1.3.1.final) in mapstruct

**Comment by filiphr** [comment]
To add to what @sjaakd said. We still have some points open about the `Before` / `After` Mapping for target objects. See https://github.com/mapstruct/mapstruct/issues/1454.

However, I am not sure that this can work properly for your case with builders and cycles

**Comment by Brutus5000** [comment]
Thanks for your answers.

No, technically it's not necessary to use builders here. I use them for convenience, to avoid unnecessary type cast with chained setters and inheritance in lombok. I guess deactivating them for MapStruct should suffice.

**Comment by thunderhook** [comment]
This was fixed with mapstruct 1.5.0, see #2721.

This is however still related with the `CycleAvoidingMappingContext` picking up the builder (instead of the resulting class). I will close this in favour of #3060.
